### PR TITLE
feat(optimizer): convert outer join to inner join

### LINKED ISSUE #7957: Feature: add optimization rule to convert outer join to inner join
**Summary**

Databend supports **left/right/full outer join**.

For some cases, we can transfer them to inner join. Such as the following case:

```sql
select * from t0 left outer join t1 on ... where t0.a > 100. 
```
If t0.a is null, then t0.a > 100 is false, so we'll filter the null rows for t0.

Finally, there are no null rows in the results, it's correct to transfer outer join to inner join.

### PR DESCRIPTION
**Pull request by xudong963**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

Speed up outer join by converting to inner join 🚀 

Fixes #7957

### CODE REVIEW

File: src/query/service/src/sql/planner/optimizer/heuristic/heuristic.rs [hunk 1]
```diff
@@ -47,6 +47,7 @@ pub static DEFAULT_REWRITE_RULES: Lazy<Vec<RuleID>> = Lazy::new(|| {
         RuleID::FoldCountAggregate,
         RuleID::SplitAggregate,
         RuleID::PushDownFilterScan,
+        RuleID::ReduceOuterJoin,
```
Comment (leiysky) [inline src/query/service/src/sql/planner/optimizer/heuristic/heuristic.rs:50]: `EliminateOuterJoin` is better.

**Comment by leiysky** [comment]
I think it's better to encapsulate this optimization in `PushDownFilterJoin`.

**Comment by xudong963** [comment]
> I think it's better to encapsulate this optimization in `PushDownFilterJoin`.

Why it's better to encapsulate this optimization in `PushDownFilterJoin`?

I think they are two independent optimization rules
1. two separated rules will be clearer and easy to debug.
2. `PushDownFilterJoin` can leverage the result of `EliminateOuterJoin` to push down predicates after left/right outer join is converted to inner join.

**Comment by leiysky** [comment]
> 1. two separated rules will be clearer and easy to debug.
> 2\. `PushDownFilterJoin` can leverage the result of `EliminateOuterJoin` to push down predicates after left/right outer join is converted to inner join.

They are actually doing the same thing, we will decide whether convert cross join to inner join or convert outer join to inner join at the same time.

And besides, the rule name `EliminateOuterJoin` can be reserved for the optimization that exploits the functional dependency of join keys to convert outer joins into inner joins.

**Comment by xudong963** [comment]
https://github.com/datafuselabs/databend/actions/runs/3157114050/jobs/5137590522 expression-related test failed @andylokandy @sundy-li

**Comment by andylokandy** [comment]
A [patch](https://github.com/datafuselabs/databend/pull/7991) is going to fix the test

**Comment by xudong963** [comment]
@mergify update

**Review by leiysky — APPROVED** [review]

**Pull request merged**